redskyctl/internal/commander/reader.go

```diff
@@ -0,0 +1,177 @@
+type ResourceReader struct {
+	Scheme           *runtime.Scheme
+	PreferredVersion runtime.GroupVersioner
+}
+
+// NewResourceReader returns a new resource reader for the supplied byte stream.
+func NewResourceReader() *ResourceReader {
+	rr := &ResourceReader{
+		Scheme:           runtime.NewScheme(),
+		PreferredVersion: OnlyVersion,
+	}
+
+	// Always add our types
+	_ = redskyv1beta1.AddToScheme(rr.Scheme)
+	_ = redskyv1alpha1.AddToScheme(rr.Scheme)
+
+	// Allow single experiments to target an experiment list
+	_ = addExperimentListConversions(rr.Scheme)
+
+	return rr
+}
+
+// ReadInto decodes the supplied byte stream into the target runtime object. The default values
+// and type information of the target object will be populated.
+func (r *ResourceReader) ReadInto(reader io.ReadCloser, target runtime.Object) error {
+	// Determine the GVK for the type we are supposed to be populating
+	gvk, err := r.objectKind(target)
+	if err != nil {
+		return err
+	}
+
+	mt := mediaType(reader)
+	decoder, err := r.decoder(mt)
+	if err != nil {
+		return err
+	}
+
+	// TODO For lists we should consider yaml.NewDocumentDecoder(reader) so we can read a stream
+
+	// Read in the raw bytes
+	defer func() { _ = reader.Close() }()
+	data, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return err
+	}
+
+	// Decode the raw data
+	obj, _, err := decoder.Decode(data, &gvk, target)
+	if err != nil {
+		return err
+	}
+
+	// If decode returned an object of a different type we should try to force the conversion
+	if obj != target {
+		if err := r.Scheme.Convert(obj, target, r.PreferredVersion); err != nil {
+			return err
+		}
+	}
+
+	// Fill in the default values for the target
+	target.GetObjectKind().SetGroupVersionKind(gvk)
+	r.Scheme.Default(target)
+
+	return nil
+}
+
+func (r *ResourceReader) objectKind(obj runtime.Object) (schema.GroupVersionKind, error) {
+	gvks, _, err := r.Scheme.ObjectKinds(obj)
+	if err != nil {
+		return schema.GroupVersionKind{}, err
+	}
+
+	gvk, ok := r.PreferredVersion.KindForGroupVersionKinds(gvks)
+	if !ok {
+		// Your code must supply a GroupVersioner to disambiguate this case
+		panic("read destination type is ambiguous in scheme")
+	}
+
+	return gvk, nil
+}
+
+func (r *ResourceReader) decoder(mediaType string) (runtime.Decoder, error) {
+	// Create the conversion serializer with our negotiation logic in it
+	cs := controller.NewConversionSerializer(r.Scheme)
+	info, ok := runtime.SerializerInfoForMediaType(cs.SupportedMediaTypes(), mediaType)
+	if !ok {
+		return nil, fmt.Errorf("could not find serializer for %s", mediaType)
+	}
+
+	return cs.DecoderToVersion(info.Serializer, r.PreferredVersion), nil
+}
+
+// mediaType returns the presumed media type for the supplied read closer.
+func mediaType(r io.ReadCloser) string {
+	// For now just assume YAML unless we got a JSON file
+	mt := runtime.ContentTypeYAML
+	if f, ok := r.(*os.File); ok {
+		switch filepath.Ext(f.Name()) {
```

Comment:
This always has bothered me for flexing on file extension. Should we try something similar to this instead?

> // YAMLOrJSONDecoder attempts to decode a stream of JSON documents or
> // YAML documents by sniffing for a leading { character.

https://github.com/kubernetes/apimachinery/blob/d4f471b82f0a17cda946aeba446770563f92114d/pkg/util/yaml/decoder.go#L326-L348

Reply:
I don't believe it is necessary to that, this is only a hint as to which parser to use. Technically JSON _is_ YAML so even if we miss the extension (e.g. `-f <(cat foo.json)`) we still end up doing the right thing (even though we don't have YAML tags on our struct fields).

I don't know what a stream of JSON documents looks like (ndjson?), but as of right now I am not using the streaming decoders (which use `ErrShortBuffer` to indicate document boundaries). We can consider adding that in later (I left a TODO), but I think it's a lot of work (e.g. we need to validate the target is a list and all documents have a valid type to go into that list). If that logic all already exists somewhere, I would be glad to use it, but that didn't fit in this timebox.